packages/opencensus-web-instrumentation-zone/src/xhr-interceptor.ts

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright 2019, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AsyncTask, XHRWithUrl } from './zone-types';
+import {
+  RootSpan,
+  Span,
+  ATTRIBUTE_HTTP_STATUS_CODE,
+  ATTRIBUTE_HTTP_METHOD,
+  parseUrl,
+  SpanKind,
+} from '@opencensus/web-core';
+import { getXhrPerfomanceData } from './perf-resource-timing-selector';
+import { traceOriginMatchesOrSameOrigin, isTrackedTask } from './util';
+import { spanContextToTraceParent } from '@opencensus/web-propagation-tracecontext';
+import {
+  annotationsForPerfTimeFields,
+  PerformanceResourceTimingExtended,
+  PERFORMANCE_ENTRY_EVENTS,
+  getResourceSpan,
+} from '@opencensus/web-instrumentation-perf';
+
+// Map intended to keep track of current XHR objects
+// associated to a span.
+const xhrSpans = new Map<XHRWithUrl, Span>();
+
+// Keeps track of the current xhr tasks that are running. This is
+// useful to clear the Performance Resource Timing entries when no
+// xhr tasks are being intercepted.
+let xhrTasksCount = 0;
+
+/**
+ * Intercepts task as XHR if it is a tracked task and its target object is
+ * instance of `XMLHttpRequest`.
+ * In case the task is intercepted, sets the Trace Context Header to it and
+ * creates a child span related to this XHR in case it is OPENED.
+ * In case the XHR is DONE, end the child span.
+ * @param task
+ */
+export function interceptXhrTask(task: AsyncTask) {
+  if (!isTrackedTask(task)) return;
+  if (!(task.target instanceof XMLHttpRequest)) return;
+
+  const xhr = task.target as XHRWithUrl;
+  if (xhr.readyState === XMLHttpRequest.OPENED) {
+    incrementXhrTaskCount();
+    const rootSpan: RootSpan = task.zone.get('data').rootSpan;
+    setTraceparentContextHeader(xhr, rootSpan);
+  } else if (xhr.readyState === XMLHttpRequest.DONE) {
+    endXhrSpan(xhr);
+    decrementXhrTaskCount();
+  }
+}
+
+function setTraceparentContextHeader(
+  xhr: XHRWithUrl,
+  rootSpan: RootSpan
+): void {
+  // `__zone_symbol__xhrURL` is set by the Zone monkey-path.
+  const xhrUrl = xhr.__zone_symbol__xhrURL;
+  const childSpan = rootSpan.startChildSpan({
+    name: parseUrl(xhrUrl).pathname,
+    kind: SpanKind.CLIENT,
+  });
+  // Associate the child span to the XHR so it allows to
+  // find the correct span when the request is DONE.
+  xhrSpans.set(xhr, childSpan);
+  if (traceOriginMatchesOrSameOrigin(xhrUrl)) {
+    xhr.setRequestHeader(
+      'traceparent',
```

Comment:
Could you make this a file level constant, e.g. `const TRACEPARENT_HEADER = 'traceparent';`? I say that because I think OpenCensus Node/Open Telemetry JS do not, mostly as a stylistic thing I think since the header itself is not likely to change!